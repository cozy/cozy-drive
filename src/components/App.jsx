import styles from '../styles/app'

import React from 'react'
import { translate } from '../plugins/preact-polyglot'
import classNames from 'classnames'

import Sidebar from './Sidebar'
import Topbar from './Topbar'

const App = ({ t, children }) => (
  <div class={classNames(styles['fil-wrapper'], styles['coz-sticky'])}>

    <Sidebar />

    <main class={styles['fil-content']}>
      <Topbar />
      { children }
    </main>
  </div>
)

export default translate()(App)
