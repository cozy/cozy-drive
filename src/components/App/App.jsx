import React from 'react'
import { Provider } from 'react-redux'
import PropTypes from 'prop-types'
import { CozyProvider } from 'cozy-client'
import { I18n } from 'cozy-ui/transpiled/react/I18n'
import { BreakpointsProvider } from 'cozy-ui/react/hooks/useBreakpoints'
import SharingProvider from 'cozy-sharing'
import { ThumbnailSizeContextProvider } from 'drive/lib/ThumbnailSizeContext'
import { ModalContextProvider } from 'drive/lib/ModalContext'

import StyledApp from 'drive/web/modules/drive/StyledApp'

const App = props => {
  return (
    <Provider store={props.store}>
      <I18n lang={props.lang} polyglot={props.polyglot}>
        <CozyProvider client={props.client}>
          <SharingProvider doctype="io.cozy.files" documentType="Files">
            <ThumbnailSizeContextProvider>
              <ModalContextProvider>
                <BreakpointsProvider>
                  <StyledApp>{props.children}</StyledApp>
                </BreakpointsProvider>
              </ModalContextProvider>
            </ThumbnailSizeContextProvider>
          </SharingProvider>
        </CozyProvider>
      </I18n>
    </Provider>
  )
}

App.propTypes = {
  store: PropTypes.object,
  lang: PropTypes.string,
  polyglot: PropTypes.object,
  client: PropTypes.object
}
export default App
