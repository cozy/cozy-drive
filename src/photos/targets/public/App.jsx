import React, { Component } from 'react'
import flatten from 'lodash/flatten'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Query } from 'cozy-client'
import { Button, Menu, MenuItem, Icon, Spinner } from 'cozy-ui/react'
import { IconSprite } from 'cozy-ui/transpiled/react'
import { Main } from 'cozy-ui/react/Layout'

import Selection from 'photos/ducks/selection'
import { MoreButton, CozyHomeLink } from 'components/Button'
import getHomeLinkHref from 'components/Button/getHomeLinkHref'
import PhotoBoard from 'photos/components/PhotoBoard'
import styles from './index.styl'
import { ALBUM_QUERY } from '../../../../src/photos/ducks/albums/index'
import ErrorUnsharedComponent from 'photos/components/ErrorUnshared'

export class App extends Component {
  onDownload = selected => {
    const photos = selected.length !== 0 ? selected : null
    this.downloadPhotos(photos)
  }

  downloadPhotos = async photos => {
    let allPhotos
    const { album } = this.props
    if (photos !== null) {
      allPhotos = flatten(photos)
    } else {
      const photosRequested = await this.context.client
        .getStackClient()
        .collection('io.cozy.files')
        .findReferencedBy(
          {
            _type: 'io.cozy.photos.albums',
            _id: album.id
          },
          { limit: 99999 }
        )
      allPhotos = photosRequested.data
    }

    this.context.client
      .collection('io.cozy.files')
      .downloadArchive(allPhotos.map(({ _id }) => _id), album.name)
  }

  renderViewer(children) {
    // children are injected by react-router when navigating a photo of the album
    if (!children) return null
    return React.Children.map(children, child =>
      React.cloneElement(child, {
        photos: this.props.photos
      })
    )
  }

  render() {
    const { album, hasMore, photos, fetchMore, children } = this.props

    const { t } = this.context
    return (
      <div
        data-test-id="pho-public-layout"
        className={styles['pho-public-layout']}
      >
        <Main className="u-pt-1-half">
          <Selection>
            {(selected, active, selection) => (
              <div>
                <div
                  className={classNames(
                    styles['pho-content-header'],
                    styles['--no-icon'],
                    styles['--hide-bar']
                  )}
                >
                  <h2 className={styles['pho-content-title']}>{album.name}</h2>
                  <div
                    data-test-id="pho-toolbar-album-public"
                    className={styles['pho-toolbar']}
                    role="toolbar"
                  >
                    <CozyHomeLink
                      from="sharing-photos"
                      t={t}
                      className={styles['pho-public-mycozy']}
                    />
                    <Button
                      theme="secondary"
                      data-test-id="album-public-download"
                      className={styles['pho-public-download']}
                      onClick={() => this.onDownload(selected)}
                      icon="download"
                      size="normal"
                      label={t('Toolbar.album_download')}
                    />

                    <Menu
                      title={t('Toolbar.more')}
                      data-test-id="more-button"
                      component={<MoreButton />}
                      position="right"
                      className="u-hide--desk"
                    >
                      <MenuItem
                        data-test-id="album-public-create-cozy-mobile"
                        onSelect={() =>
                          (window.location = getHomeLinkHref('sharing-photos'))
                        }
                        icon={<Icon icon="cozy-negative" />}
                      >
                        {t('Share.create-cozy')}
                      </MenuItem>
                      <MenuItem
                        data-test-id="album-public-download-mobile"
                        onSelect={() => this.onDownload(selected)}
                        icon={<Icon icon="download" />}
                      >
                        {t('Toolbar.album_download')}
                      </MenuItem>
                    </Menu>
                  </div>
                </div>
                <PhotoBoard
                  lists={[{ photos }]}
                  selected={selected}
                  photosContext="timeline"
                  showSelection={active}
                  onPhotoToggle={selection.toggle}
                  onPhotosSelect={selection.select}
                  onPhotosUnselect={selection.unselect}
                  fetchStatus={photos.fetchStatus}
                  hasMore={hasMore}
                  fetchMore={fetchMore}
                />
                {this.renderViewer(children)}
              </div>
            )}
          </Selection>
          <IconSprite />
        </Main>
      </div>
    )
  }
}
App.propTypes = {
  album: PropTypes.object.isRequired,
  hasMore: PropTypes.bool, //see https://github.com/cozy/cozy-client/issues/345
  photos: PropTypes.array.isRequired,
  fetchMore: PropTypes.func.isRequired
}
App.contextTypes = {
  t: PropTypes.func
}
const ConnectedApp = props => (
  <Query query={ALBUM_QUERY} {...props}>
    {({ data: album, fetchStatus }) => {
      if (fetchStatus === 'failed') {
        return <ErrorUnsharedComponent />
      }
      if (fetchStatus === 'loaded') {
        return (
          <App
            album={album}
            photos={album.photos.data}
            hasMore={album.photos.hasMore}
            fetchMore={album.photos.fetchMore.bind(album.photos)}
          >
            {props.children}
          </App>
        )
      } else {
        return (
          <Spinner
            size={'xxlarge'}
            loadingType={'photos_fetching'}
            middle={true}
          />
        )
      }
    }}
  </Query>
)

export default ConnectedApp
