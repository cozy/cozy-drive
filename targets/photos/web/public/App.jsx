import React, { Component } from 'react'
import { Query } from 'cozy-client'

import Selection from 'photos/ducks/selection'
import PhotoBoard from 'photos/components/PhotoBoard'
import Loading from 'photos/components/Loading'
import ErrorShare from 'components/Error/ErrorShare'
import { Button } from 'cozy-ui/react'
import { MoreButton, CozyHomeLink } from 'components/Button'
import Menu, { Item } from 'components/Menu'

import classNames from 'classnames'
import styles from './index.styl'

class App extends Component {
  onDownload = selected => {
    const photos =
      selected.length !== 0 ? selected : this.props.album.photos.data
    this.downloadPhotos(photos)
  }

  downloadPhotos(photos) {
    this.context.client
      .collection('io.cozy.files')
      .downloadArchive(photos.map(({ _id }) => _id), this.props.album.name)
  }

  componentWillReceiveProps(nextProps) {
    const { fetchStatus } = nextProps
    if (fetchStatus === 'failed') {
      this.setState({ error: 'Fetch error' })
    }
  }

  render() {
    const { album, fetchStatus } = this.props

    const { t } = this.context
    if (this.state.error) {
      return (
        <div
          className={classNames(
            styles['pho-public-layout'],
            styles['pho-public-layout--full']
          )}
        >
          <ErrorShare errorType={`public_album_unshared`} />
        </div>
      )
    }
    if (fetchStatus === 'pending' || fetchStatus === 'loading') {
      return (
        <div className={styles['pho-public-layout']}>
          <Loading loadingType="photos_fetching" />
        </div>
      )
    }
    const { data, hasMore } = album.photos
    return (
      <div className={styles['pho-public-layout']}>
        <Selection>
          {(selected, active, selection) => (
            <div>
              <div
                className={classNames(
                  styles['pho-content-header'],
                  styles['--no-icon']
                )}
              >
                <h2 className={styles['pho-content-title']}>{album.name}</h2>
                <div className={styles['pho-toolbar']} role="toolbar">
                  <Button
                    theme="secondary"
                    className={styles['pho-public-download']}
                    onClick={() => this.onDownload(selected)}
                    icon="download"
                    label={t('Toolbar.album_download')}
                  />
                  <CozyHomeLink from="sharing-photos" />
                  <Menu
                    title={t('Toolbar.more')}
                    className={classNames(styles['pho-toolbar-menu'])}
                    button={<MoreButton>{t('Toolbar.more')}</MoreButton>}
                  >
                    <Item>
                      <a
                        className={classNames(styles['pho-public-download'])}
                        onClick={() => this.onDownload(selected)}
                      >
                        {t('Toolbar.album_download')}
                      </a>
                    </Item>
                  </Menu>
                </div>
              </div>
              <PhotoBoard
                photosContext="shared_album"
                lists={[{ photos: data }]}
                selected={selected}
                showSelection={active}
                onPhotoToggle={selection.toggle}
                onPhotosSelect={selection.select}
                onPhotosUnselect={selection.unselect}
                hasMore={hasMore}
                fetchMore={() => album.photos.fetchMore()}
              />
              {this.renderViewer(this.props.children)}
            </div>
          )}
        </Selection>
      </div>
    )
  }

  renderViewer(children) {
    if (!children) return null
    return React.Children.map(children, child =>
      React.cloneElement(child, {
        photos: this.props.album.photos.data
      })
    )
  }
}

const ConnectedApp = props => (
  <Query
    query={client =>
      client
        .get('io.cozy.photos.albums', props.router.params.albumId)
        .include(['photos'])
    }
  >
    {({ data, fetchStatus }) => (
      <App album={data || null} fetchStatus={fetchStatus} {...props} />
    )}
  </Query>
)

export default ConnectedApp
