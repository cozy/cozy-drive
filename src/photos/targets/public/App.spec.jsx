import React from 'react'
import { shallow } from 'enzyme'
import { App } from './App'

describe('Public view', () => {
  const options = {
    context: {
      t: jest.fn()
    }
  }

  it('should render the album', () => {
    const app = shallow(
      <App album={{}} hasMore={false} fetchMore={jest.fn()} photos={[]} />,
      options
    )
    const selection = app.find('Selection')
    expect(selection.dive()).toMatchSnapshot()
  })

  it('should render children when they are present', () => {
    const app = shallow(
      <App album={{}} hasMore={false} fetchMore={jest.fn()} photos={[1, 2, 3]}>
        <div />
      </App>,
      options
    )
    const selection = app.find('Selection')
    expect(selection.dive()).toMatchSnapshot()
  })
})
