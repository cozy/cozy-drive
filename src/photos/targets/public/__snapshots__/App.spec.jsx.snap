// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Public view should render children when they are present 1`] = `
<Provider
  value={Array []}
>
  <div>
    <div>
      <div
        className=""
      >
        <h2 />
        <div
          data-test-id="pho-toolbar-album-public"
          role="toolbar"
        >
          <CozyHomeLink
            from="sharing-photos"
            size="normal"
            t={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "Toolbar.album_download",
                  ],
                  Array [
                    "Toolbar.more",
                  ],
                  Array [
                    "Share.create-cozy",
                  ],
                  Array [
                    "Toolbar.album_download",
                  ],
                  Array [
                    "Toolbar.album_download",
                  ],
                  Array [
                    "Toolbar.more",
                  ],
                  Array [
                    "Share.create-cozy",
                  ],
                  Array [
                    "Toolbar.album_download",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
          />
          <Button
            data-test-id="album-public-download"
            icon="download"
            onClick={[Function]}
            size="normal"
            tag="button"
            theme="secondary"
            type="submit"
          />
          <Menu
            className="u-hide--desk"
            component={<Wrapper />}
            data-test-id="more-button"
            disabled={false}
            itemsStyle={Object {}}
            onSelect={null}
            onSelectDisabled={null}
            popover={false}
            position="right"
          >
            <MenuItem
              data-test-id="album-public-create-cozy-mobile"
              icon={
                <Icon
                  icon="cozy-negative"
                  spin={false}
                />
              }
              onSelect={[Function]}
            />
            <MenuItem
              data-test-id="album-public-download-mobile"
              icon={
                <Icon
                  icon="download"
                  spin={false}
                />
              }
              onSelect={[Function]}
            />
          </Menu>
        </div>
      </div>
      <Wrapper
        fetchMore={[MockFunction]}
        hasMore={false}
        lists={
          Array [
            Object {
              "photos": Array [
                1,
                2,
                3,
              ],
            },
          ]
        }
        onPhotoToggle={[Function]}
        onPhotosSelect={[Function]}
        onPhotosUnselect={[Function]}
        photosContext="timeline"
        selected={Array []}
        showSelection={false}
      />
      <div
        key=".0"
        photos={
          Array [
            1,
            2,
            3,
          ]
        }
      />
    </div>
  </div>
</Provider>
`;

exports[`Public view should render the album 1`] = `
<Provider
  value={Array []}
>
  <div>
    <div>
      <div
        className=""
      >
        <h2 />
        <div
          data-test-id="pho-toolbar-album-public"
          role="toolbar"
        >
          <CozyHomeLink
            from="sharing-photos"
            size="normal"
            t={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "Toolbar.album_download",
                  ],
                  Array [
                    "Toolbar.more",
                  ],
                  Array [
                    "Share.create-cozy",
                  ],
                  Array [
                    "Toolbar.album_download",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
          />
          <Button
            data-test-id="album-public-download"
            icon="download"
            onClick={[Function]}
            size="normal"
            tag="button"
            theme="secondary"
            type="submit"
          />
          <Menu
            className="u-hide--desk"
            component={<Wrapper />}
            data-test-id="more-button"
            disabled={false}
            itemsStyle={Object {}}
            onSelect={null}
            onSelectDisabled={null}
            popover={false}
            position="right"
          >
            <MenuItem
              data-test-id="album-public-create-cozy-mobile"
              icon={
                <Icon
                  icon="cozy-negative"
                  spin={false}
                />
              }
              onSelect={[Function]}
            />
            <MenuItem
              data-test-id="album-public-download-mobile"
              icon={
                <Icon
                  icon="download"
                  spin={false}
                />
              }
              onSelect={[Function]}
            />
          </Menu>
        </div>
      </div>
      <Wrapper
        fetchMore={[MockFunction]}
        hasMore={false}
        lists={
          Array [
            Object {
              "photos": Array [],
            },
          ]
        }
        onPhotoToggle={[Function]}
        onPhotosSelect={[Function]}
        onPhotosUnselect={[Function]}
        photosContext="timeline"
        selected={Array []}
        showSelection={false}
      />
    </div>
  </div>
</Provider>
`;
